implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/DeltaOtlpMeterRegistryTest.java

```diff
@@ -0,0 +1,325 @@
+                .register(registry);
+        counter.increment();
+        counter.increment();
+        assertSum(publishTimeAwareWrite(counter), 0, TimeUnit.MINUTES.toNanos(1), 0);
+        this.stepOverNStep(1);
+        assertSum(publishTimeAwareWrite(counter), TimeUnit.MINUTES.toNanos(1), TimeUnit.MINUTES.toNanos(2), 2);
+
+        this.stepOverNStep(1);
+        counter.increment();
+        assertSum(publishTimeAwareWrite(counter), TimeUnit.MINUTES.toNanos(2), TimeUnit.MINUTES.toNanos(3), 1);
+    }
+
+    @Test
+    void functionCounter() {
+        AtomicLong atomicLong = new AtomicLong(10);
+
+        FunctionCounter counter = FunctionCounter.builder(METER_NAME, atomicLong, AtomicLong::get)
+                .description(METER_DESCRIPTION).tags(Tags.of(meterTag)).baseUnit("milliseconds").register(registry);
+
+        assertSum(publishTimeAwareWrite(counter), 0, TimeUnit.MINUTES.toNanos(1), 0);
+        this.stepOverNStep(1);
+        assertSum(publishTimeAwareWrite(counter), TimeUnit.MINUTES.toNanos(1), TimeUnit.MINUTES.toNanos(2), 10);
+        this.stepOverNStep(1);
+        assertSum(publishTimeAwareWrite(counter), TimeUnit.MINUTES.toNanos(2), TimeUnit.MINUTES.toNanos(3), 0);
+    }
+
+    @Test
+    void timer() {
+        Timer timer = Timer.builder(METER_NAME).description(METER_DESCRIPTION).tags(Tags.of(meterTag))
+                .register(registry);
+        timer.record(10, TimeUnit.MILLISECONDS);
+        timer.record(77, TimeUnit.MILLISECONDS);
+        timer.record(111, TimeUnit.MILLISECONDS);
+
+        // This is where TimeWindowMax can be painful and make no sense at all. Need to
+        // re-visit this. Probably StepMax is what might fit good for OTLP.
+        assertHistogram(publishTimeAwareWrite(timer), 0, TimeUnit.MINUTES.toNanos(1), "milliseconds", 0, 0, 111);
+        this.stepOverNStep(1);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(1), TimeUnit.MINUTES.toNanos(2),
+                "milliseconds", 3, 198, 111);
+        timer.record(4, TimeUnit.MILLISECONDS);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(1), TimeUnit.MINUTES.toNanos(2),
+                "milliseconds", 3, 198, 111);
+        this.stepOverNStep(1);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(2), TimeUnit.MINUTES.toNanos(3),
+                "milliseconds", 1, 4, 111);
+
+        this.stepOverNStep(2);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(4), TimeUnit.MINUTES.toNanos(5),
+                "milliseconds", 0, 0, 0);
+        timer.record(1, TimeUnit.MILLISECONDS);
+        this.stepOverNStep(1);
+        assertHistogram(publishTimeAwareWrite(timer), TimeUnit.MINUTES.toNanos(5), TimeUnit.MINUTES.toNanos(6),
+                "milliseconds", 1, 1, 1);
+    }
+
+    @Test
+    void timerWithHistogram() {
```

Comment:
@shakuzen / @jonatan-ivanov I believe you guys are already aware that the TImeWindowFixedBoundaryHistogram will have very poor accuracy for StepBased registries. 

To Summarise:
The issue with that is it uses a sliding bucket which creates 3 buckets(assumption, step=1min, the default buffer length is 3). And these buckets are rotated every 20 seconds (step/buffer length) which makes the histogram always error-prone I,e it will always undercount the values. This is the reason we used [DeltaHistogramCounts](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-signalfx/src/main/java/io/micrometer/signalfx/DeltaHistogramCounts.java) implementation in SignalFxMeterRegistry for our histogram use-cases. With OTLP coming into the picture, I think we should at least have a similar implementation (or basic StepBased Histogram) which will give more accurate information. And moreover, the percentiles [don't align with Step Boundaries ](https://github.com/micrometer-metrics/micrometer/issues/3087) which also makes it difficult to correlate them.

Reply:
We'll need to change this. SignalFx was the first step-based registry to support percentile histograms, so we hadn't taken close enough look at this in general, I think. The histogram support is indeed not right, which we can fix after merging this, but I don't usually like merging something I know has a bug.

The percentiles and max not aligning with step boundaries was intentional (there's a relevant old issue with the history), but I don't think it makes sense in the OTLP context because the specification is clear about what the meaning of the max or bucket counts is supposed to represent, and it needs to align with the time interval.